src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -72,32 +73,89 @@ nodeExtractText = T.strip . mconcat . map extractText . nodeFlatten
     nodeFlatten :: Node -> [NodeType]
     nodeFlatten (Node _pos ty subs) = ty : concatMap nodeFlatten subs
 
+
 data IgnoreMode
-  = Link
-  | Paragraph
-  | File
+  = IMLink
+  | IMParagraph
+  | IMFile
+  deriving stock (Eq)
+
+-- | "ignore link" pragmas in different places behave slightly different,
+-- so @IgnoreMode@ @Link@ is parametrized
+data IgnoreLinkState
+  = StatusNotSetYet
+  -- ^ @getIgnoreMode@ doesn't know context, so we'll need to set it later
+  | ExpectingLinkInParagraph
+  -- ^ When ignore annotation is inside @PARAGRAPH@ node,
+  -- we expect a link to ignore later in this paragraph.
+  -- We raise scan error if we see this status after
+  -- traversing subnodes of a @PARAGRAPH@ node.
+  | ExpectingLinkInSubnodes
+  -- ^ If ignore annotation is not inside @PARAGRAPH@, then we expect a link
+  -- in subtree of next node. We raise scan error if we see this status
+  -- after traversing childs of any node that is not an ignore annotation.
+  | ParentExpectsLink
+  -- ^ When we have `WaitingLinkInSubnodes`, we traverse subtree of some node,
+  -- and we should change `IgnoreLinkStatus`, because it's not a problem if
+  -- our node's first child doesn't contain a link. So this status means that
+  -- we won't throw errors if we don't find a link for now
+  deriving stock (Eq)
+
+data IgnoreModeState
+  = IMSLink IgnoreLinkState
+  | IMSParagraph
+  | IMSFile
   deriving stock (Eq)
 
 -- | Bind `IgnoreMode` to its `PosInfo` so that we can tell where the
 -- corresponding annotation was declared.
-data Ignore = Ignore IgnoreMode (Maybe PosInfo)
+data Ignore = Ignore
+  { _ignoreMode :: IgnoreModeState
+  , _ignorePos :: Maybe PosInfo
+  }
+makeLensesFor [("_ignoreMode", "ignoreMode")] 'Ignore
 
 data GetIgnoreMode
   = NotAnAnnotation
   | ValidMode IgnoreMode
   | InvalidMode Text
   deriving stock (Eq)
 
-type ScannerM a = StateT (Maybe Ignore) (Writer [ScanError]) a
+
+
+data ScannerState = ScannerState
+  { _ssIgnore :: Maybe Ignore
+  , _ssParentNodeType :: Maybe NodeType
+  -- ^ @cataNodeWithParentNodeInfo@ allows to get a @NodeType@ of parent node from this field
+  }
+makeLenses ''ScannerState
+
+initialScannerState :: ScannerState
+initialScannerState = ScannerState
+  { _ssIgnore = Nothing
+  , _ssParentNodeType = Nothing
+  }
+
+type ScannerM a = StateT ScannerState (Writer [ScanError]) a
 
 -- | A fold over a `Node`.
 cataNode :: (Maybe PosInfo -> NodeType -> [c] -> c) -> Node -> c
 cataNode f (Node pos ty subs) = f pos ty (cataNode f <$> subs)
 
+-- | Sets correct @_ssParentNodeType@ before running scanner on each node
+cataNodeWithParentNodeInfo
+  :: (Maybe PosInfo -> NodeType -> [ScannerM a] -> ScannerM a)
+  -> Node
+  -> ScannerM a
+cataNodeWithParentNodeInfo f node = cataNode f' node
+  where
+    f' pos ty childScanners = f pos ty $
+      map (ssParentNodeType .= Just ty >>) childScanners
+
 -- | Remove nodes with accordance with global `MarkdownConfig` and local
 --   overrides.
```

Comment:
I just noticed this old comment here, it doesn't look exactly right, does it?

We don't use `MarkdownConfig` at all here. We only care about `<!-- xrefcheck: ignore <mode> -->` annotations :thinking: 

Can you please rewrite this haddock?

Reply:
Done